File: android/core/protocol/src/main/kotlin/fr/mina/gateway/protocol/MediaChunker.kt
package fr.mina.gateway.protocol

import org.json.JSONObject
import java.security.MessageDigest
import java.security.SecureRandom

/**
 * Découpe un média (image recompressée / audio m4a) en un payload de métadonnées + N payloads de
 * chunks binaires, encodés en payload v2 (ChatPayloadCodec). Miroir de src/chat/media-chunker.mjs :
 * le PC réassemble avec les mêmes bornes et vérifie le sha256. 128 Kio par chunk (sous le plafond
 * ciphertext de l'enveloppe), 5 Mo max au total.
 */
object MediaChunker {
    private const val CHUNK_BYTES = 131_072
    private const val MAX_TOTAL_BYTES = 5 * 1024 * 1024
    private const val MAX_CHUNK_COUNT = 4_096
    private val MEDIA_ID = Regex("^[A-Za-z0-9._:-]{1,64}$")
    private val SHA256 = Regex("^[a-f0-9]{64}$")
    private val MIME_TO_TYPE = mapOf(
        "image/jpeg" to "message.attachment.created",
        "image/png" to "message.attachment.created",
        "image/webp" to "message.attachment.created",
        "audio/mp4" to "message.voice.created",
        VoicePcmFormat.MIME to "message.voice.created",
    )

    data class MediaPiece(val mediaId: String, val metaPayload: ByteArray, val chunkPayloads: List<ByteArray>) {
        override fun equals(other: Any?): Boolean = this === other
        override fun hashCode(): Int = mediaId.hashCode()
    }

    private fun randomMediaId(): String {
        val bytes = ByteArray(16).also { SecureRandom().nextBytes(it) }
        return bytes.joinToString("") { "%02x".format(it) }
    }

    private fun sha256Hex(bytes: ByteArray): String =
        MessageDigest.getInstance("SHA-256").digest(bytes).joinToString("") { "%02x".format(it) }

    fun eventTypeFor(mime: String): String =
        MIME_TO_TYPE[mime] ?: throw IllegalArgumentException("media_mime_refuse:$mime")

    fun encodeMeta(
        mediaId: String,
        mime: String,
        sizeBytes: Int,
        sha256: String,
        chunkCount: Int,
        chunkBytes: Int,
        extraMeta: Map<String, Any> = emptyMap(),
    ): ByteArray {
        val eventType = eventTypeFor(mime)
        val maxBytes = maxBytesFor(mime)
        val maxChunkBytes = maxChunkBytesFor(mime)
        require(MEDIA_ID.matches(mediaId)) { "media_id_invalide" }
        require(sizeBytes in 1..maxBytes) { "media_taille_invalide" }
        require(SHA256.matches(sha256)) { "media_digest_invalide" }
        require(chunkBytes in 1..maxChunkBytes) { "media_chunk_bytes_invalide" }
        require(chunkCount in 1..MAX_CHUNK_COUNT) { "media_chunk_count_invalide" }
        val theoreticalBytes = chunkCount.toLong() * chunkBytes.toLong()
        val theoreticalLimit = if (VoicePcmFormat.isCanonicalMime(mime)) {
            maxBytes.toLong() + VoicePcmFormat.CHUNK_BYTES - 1
        } else {
            maxBytes.toLong()
        }
        require(theoreticalBytes <= theoreticalLimit) { "media_trop_gros" }
        require(chunkCount == (sizeBytes + chunkBytes - 1) / chunkBytes) { "media_chunk_count_invalide" }
        return ChatPayloadCodec.encodeV2(
            eventType,
            JSONObject().apply {
                put("mediaId", mediaId)
                put("mime", mime)
                put("sizeBytes", sizeBytes)
                put("sha256", sha256)
                put("chunkCount", chunkCount)
                put("chunkBytes", chunkBytes)
                for ((key, value) in extraMeta) put(key, value)
            }.toString(),
        )
    }

    fun encodeChunk(mediaId: String, index: Int, binary: ByteArray): ByteArray {
        require(MEDIA_ID.matches(mediaId)) { "media_id_invalide" }
        require(index >= 0) { "media_index_invalide" }
        require(binary.isNotEmpty() && binary.size <= CHUNK_BYTES) { "media_chunk_taille_invalide" }
        return ChatPayloadCodec.encodeV2(
            "media.chunk",
            JSONObject().apply {
                put("mediaId", mediaId)
                put("index", index)
            }.toString(),
            binary,
        )
    }

    fun chunk(bytes: ByteArray, mime: String, extraMeta: Map<String, Any> = emptyMap()): MediaPiece {
        eventTypeFor(mime)
        require(bytes.isNotEmpty()) { "media_vide" }
        val maxBytes = maxBytesFor(mime)
        val chunkBytes = maxChunkBytesFor(mime)
        require(bytes.size <= maxBytes) { "media_trop_gros" }

        val mediaId = randomMediaId()
        val chunkCount = (bytes.size + chunkBytes - 1) / chunkBytes
        val metaPayload = encodeMeta(
            mediaId = mediaId,
            mime = mime,
            sizeBytes = bytes.size,
            sha256 = sha256Hex(bytes),
            chunkCount = chunkCount,
            chunkBytes = chunkBytes,
            extraMeta = extraMeta,
        )

        val chunkPayloads = ArrayList<ByteArray>(chunkCount)
        for (index in 0 until chunkCount) {
            val from = index * chunkBytes
            val to = minOf(from + chunkBytes, bytes.size)
            val slice = bytes.copyOfRange(from, to)
            chunkPayloads.add(encodeChunk(mediaId, index, slice))
        }
        return MediaPiece(mediaId, metaPayload, chunkPayloads)
    }

    private fun maxBytesFor(mime: String): Int =
        if (VoicePcmFormat.isCanonicalMime(mime)) VoicePcmFormat.MAX_BYTES else MAX_TOTAL_BYTES

    private fun maxChunkBytesFor(mime: String): Int =
        if (VoicePcmFormat.isCanonicalMime(mime)) VoicePcmFormat.CHUNK_BYTES else CHUNK_BYTES
}

File: src/chat/media-chunker.mjs
import { createHash, randomBytes } from 'node:crypto';
import { encodeChatPayloadV2 } from '../contracts/chat-payload.mjs';
import {
  VOICE_CHUNK_BYTES,
  VOICE_MAX_BYTES,
  VOICE_PCM_MIME,
  isVoicePcmMime,
} from './voice-pcm.mjs';

// Découpe un média (déjà normalisé : image recompressée / audio m4a) en un événement de métadonnées
// + N chunks binaires, prêts à chiffrer et envoyer sur le canal `mina_app`. Symétrique exact du
// réassembleur (media-assembler) : chunk() puis assemble() redonne les octets d'origine. Sert au
// sens PC → téléphone (W6) et de référence pour le chunker Android (W3).

const DEFAULT_CHUNK_BYTES = 131_072; // 128 Kio (sous le plafond ciphertext de l'enveloppe)
const DEFAULT_MAX_TOTAL_BYTES = 5 * 1024 * 1024;
const MIME_TO_TYPE = Object.freeze({
  'image/jpeg': 'message.attachment.created',
  'image/png': 'message.attachment.created',
  'image/webp': 'message.attachment.created',
  'audio/mp4': 'message.voice.created',
  [VOICE_PCM_MIME]: 'message.voice.created',
});

const newMediaId = () => randomBytes(16).toString('hex'); // 32 hex, conforme à MEDIA_ID de l'assembleur

/**
 * @param {Buffer} bytes          contenu déjà normalisé
 * @param {object} opts
 * @param {string} opts.mime      un mime de MIME_TO_TYPE
 * @param {number} [opts.chunkBytes]
 * @param {number} [opts.maxTotalBytes]
 * @param {object} [opts.extraMeta]  ex. { width, height } pour une image, { durationMs } pour l'audio
 * @param {Function} [opts.makeId]
 * @returns {{ eventType, meta, chunks:Array<{index:number, binary:Buffer}> }}
 */
export function chunkMedia(bytes, {
  mime,
  chunkBytes = null,
  maxTotalBytes = null,
  extraMeta = {},
  makeId = newMediaId,
} = {}) {
  const buffer = Buffer.from(bytes ?? []);
  const eventType = MIME_TO_TYPE[mime];
  const voicePcm = isVoicePcmMime(mime);
  const resolvedChunkBytes = chunkBytes ?? (voicePcm ? VOICE_CHUNK_BYTES : DEFAULT_CHUNK_BYTES);
  const requestedMaxTotalBytes = maxTotalBytes ?? (voicePcm ? VOICE_MAX_BYTES : DEFAULT_MAX_TOTAL_BYTES);
  const resolvedMaxTotalBytes = voicePcm
    ? Math.min(requestedMaxTotalBytes, VOICE_MAX_BYTES)
    : requestedMaxTotalBytes;
  if (!eventType) throw new Error(`media_mime_refuse:${mime}`);
  if (buffer.length < 1) throw new Error('media_vide');
  if (buffer.length > resolvedMaxTotalBytes) throw new Error('media_trop_gros');
  if (!Number.isInteger(resolvedChunkBytes) || resolvedChunkBytes < 1 || resolvedChunkBytes > (voicePcm ? VOICE_CHUNK_BYTES : DEFAULT_CHUNK_BYTES)) throw new Error('media_chunk_bytes_invalide');

  const mediaId = makeId();
  const chunkCount = Math.ceil(buffer.length / resolvedChunkBytes);
  const chunks = [];
  for (let index = 0; index < chunkCount; index += 1) {
    chunks.push({ index, binary: Buffer.from(buffer.subarray(index * resolvedChunkBytes, (index + 1) * resolvedChunkBytes)) });
  }
  const meta = Object.freeze({
    mediaId,
    mime,
    sizeBytes: buffer.length,
    sha256: createHash('sha256').update(buffer).digest('hex'),
    chunkCount,
    chunkBytes: resolvedChunkBytes,
    ...extraMeta,
  });
  return Object.freeze({ eventType, meta, chunks: Object.freeze(chunks) });
}

/** Encode le payload v2 de l'événement métadonnées (à chiffrer ensuite). */
export function encodeMediaMetaPayload({ eventType, meta }) {
  return encodeChatPayloadV2({ type: eventType, meta });
}

/** Encode le payload v2 d'un chunk (à chiffrer ensuite). */
export function encodeMediaChunkPayload({ mediaId, index, binary }) {
  return encodeChatPayloadV2({ type: 'media.chunk', meta: { mediaId, index }, binary });
}

export { DEFAULT_CHUNK_BYTES };
